### PR TITLE
Mnaveau/free flyer and com torque

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

[//]: # "Description of what you did.  If it is more complex, consider to add a"
[//]: # "'Summary' section."

- Add a `ImpedanceController::get_joint_torques()` method.

## How I Tested

[//]: # "Explain how you tested your changes"

Ran simu on pybullet.

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [ ] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [ ] All new functions/classes are documented and existing documentation is updated according to changes.
- [ ] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

File: src/centroidal_force_qp_controller.cpp [hunk 1]
```diff
@@ -103,7 +103,8 @@ void CentroidalForceQPController::run(
     {
         double x = relative_position_endeff(3 * i);
         double y = relative_position_endeff(3 * i + 1);
-        double z = 0;  // Always assumed to be on the floor.
+        // double z = 0;  // Always assumed to be on the floor.
+        double z = relative_position_endeff(3 * i + 2);
```
Comment (jviereck) [inline src/centroidal_force_qp_controller.cpp:107]: Can you remove the comment here?

File: src/impedance_controller.cpp [hunk 2]
```diff
@@ -34,14 +34,23 @@ void ImpedanceController::initialize(const pinocchio::Model& pinocchio_model,
...
     end_jacobian_.resize(6, pinocchio_model_.nv);
     end_jacobian_.fill(0.);
-    root_jacobian_.fill(0.);
-
     impedance_jacobian_.resize(6, pinocchio_model_.nv);
+    impedance_jacobian_.fill(0.);
+
+    // Defines if the model has a freeflyer.
+    pinocchio_model_has_free_flyer_ =
+        pinocchio_model_.joints[1].shortname() == "JointModelFreeFlyer";
```
Comment (jviereck) [inline src/impedance_controller.cpp:45]: I like this one :)!

**Review by jviereck — CHANGES_REQUESTED** [review]
Looks good overall with small request.

Are you sure changing the reference frame for the jacobians from LOCAL_WORLD_ALIGNED -> WORLD is the right thing to do? It was working fine before with LOCAL_WORLD_ALIGNED for me.

**Comment by jviereck** [comment]
Trying to compile this PR, I am getting quite a few errors like this one:

```
/home/jviereck/dev/blmc_testing/workspace/src/mim_control/src/impedance_controller.cpp:130:49: error: no match for ‘operator*’ (operand types are ‘Eigen::Transpose<Eigen::Matrix<double, 6, -1, 0, 6, -1> >’ and ‘pinocchio::Force {aka pinocchio::ForceTpl<double, 0>}’)
     torques_ = (impedance_jacobian_.transpose() * impedance_force_);
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
```

They seem all related to the `impedance_force_` being a `pinocchio::Force` instead of a 6-dim-vector.

**Comment by jviereck** [comment]
I switched the `impedance_force_` from a `pinocchio::Force` to a `Vector6d`. With this change the code is compiling. However, the robot starts flying away when I try to walk with the changes in simulation.

**Comment by MaximilienNaveau** [comment]
Sorry this PR is under debug

**Comment by MaximilienNaveau** [comment]
> Looks good overall with small request.
> 
> Are you sure changing the reference frame for the jacobians from LOCAL_WORLD_ALIGNED -> WORLD is the right thing to do? It was working fine before with LOCAL_WORLD_ALIGNED for me.

I think you did try only using root_frame as "universe" and not the hip or something.
And indeed the code should work for this case but I think there a bug if you try a subkinematic chain

**Comment by MaximilienNaveau** [comment]
@jviereck I force pushed the PR I just add the get_joint_torques.
There is still an issue if I try to control the foot in the hip frame...
Maybe we should chat about this

**Review by jviereck — APPROVED** [review]
This looks good.

**Pull request merged**